Check term format

pranitbauva1997 opened the pull request:
This is RFC. Changes will be made according to the reviews

pranitbauva1997 commented:
@chriscool @larsxschneider I was able to fix the previous bug in my branch check-term-format regarding the test no 68 by using die(). When I used your suggestion to use "return error()" instead of a normal "printf", what it does is that it prints it to the stderr instead of stdout. This creates a problem in test no 46 and 47 in t/t6030 as it redirects the stderr stream to the file and thus there is a miss match. Though using error() is more appropriate, it seems as if the current behavior does not reflect it. I then reverted back to using printf() and the whole test suite passes. What should I do in this case?

pranitbauva1997 commented on git-bisect.sh:
```diff
@@ -564,15 +564,29 @@ write_terms () {
+	echo "TERM_BAD: $TERM_BAD"
+	echo "TERM_GOOD: $TERM_GOOD"
+	git bisect--helper --check-term-format "$TERM_BAD" bad
+	if [ $? -eq 128 ]; then
```
If you are wondering why I have done this, then it is because "die()" does not work perfectly when switch from shell to C back and forth. When I use die() in C, it completely dies out in the C code but it has to be explicitly specified to die in the shell script also.

larsxschneider replied:
alternatively you could do this (which would die if there is any return non-equal 0):
```sh
set -e
git-bisect--helper ...
git-bisect--helper ...
set +e
```
However, I am not sure if this is OK regarding Git coding conventions. @chriscool do you know?

pranitbauva1997 replied:
This is the biggest problem I faced. The error() function also returns a -1 value but the program is not supposed to terminate there. I initially tried this but it failed and it took me quite a lot of time to figure this stuff out. What I believe is that since the modifications made in git-bisect.sh are temporary which would be soon replaced, it would be OKAY.

chriscool replied:
We don't use "set -e" and stuff like that in Git. I think that's because it is not easily portable and has edge cases.

chriscool replied:
Did you try something like:

```
git bisect--helper --check-term-format "$TERM_BAD" bad || exit
git bisect--helper --check-term-format "$TERM_GOOD" good || exit
```
?

In which case is it not working?

pranitbauva1997 replied:
This set was my first intuition which failed a lot of tests.

> git bisect--helper --check-term-format "$TERM_BAD" bad || exit
git bisect--helper --check-term-format "$TERM_GOOD" good || exit

So then after spending a lot of time, I figured out the current behavior. The reason why your suggested code will fail is because even the error() return -1 value and thus your implementation will make the program exit but it is not supposed to exit at that time.

chriscool commented:
It looks like you didn't force push to this repo the changes you made in your branch, right?
So how can we see these changes?

It's much easier to comment on some changes when we can see them, so please show us your changes.

pranitbauva1997 commented:
@chriscool I have made another commit which shows the changes which can be introduced in order to make the whole test suite pass.

larsxschneider commented on git-bisect.sh:
```diff
@@ -564,15 +564,29 @@ write_terms () {
+	if [ $? -eq 128 ]; then
+		die "bye"
+	fi
+	git bisect--helper --check-term-format "$TERM_GOOD" good
```
Some scripts seem to use `git-foo` and others use `git foo`. I guess it doesn't matter that much but the former might be (measurable?) faster... especially on Windows very process creation takes ages...
@chriscool: do you know the coding conventions here?

chriscool replied:
I think it's better to use `git foo` everywhere. I think if the `git` binary is already in memory, it may avoid loading the `git-foo` binary if the foo command is implemented in C. (And yeah, some commands are sometimes converted to C :-) Not sure about Windows, but we should not change things to make it faster for Windows, if it could makes it slower for other OSes.

larsxschneider commented on builtin/bisect--helper.c:
```diff
@@ -2,19 +2,75 @@
+	va_start(matches, term);
+	while((match = va_arg(matches, const char *)) != NULL)
+		if(!strcmp(term, match))
+			return 1;
```
I personally dislike omitting braces because it makes it harder to parse for me... however, I think this is Git coding style 😄

chriscool replied:
Yeah, but there should be a space between `if` and `(` and also between `while` and `(`.

pranitbauva1997 replied:
I will take care about the little things a bit later. As for now, I am focusing on the bigger picture of copying the behavior of function.

larsxschneider commented on builtin/bisect--helper.c:
```diff
@@ -2,19 +2,75 @@
+	CHECK_TERM_FMT
+};
+
+static int one_of(const char *term, ...)
```
can you find a more descriptive function name here? or add a one sentence comment what the function does?

pranitbauva1997 replied:
I will add a comment here.

chriscool replied:
I think one_of() was suggested by Junio, so it probably ok. But yeah a comment might be nice.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -2,19 +2,75 @@
+	char new_term[1000] = "refs/bisect/";
+	strcat(new_term, term);
+
+	if (check_refname_format(new_term, flag)) {
```
In this function the `if`s are ok.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -2,19 +2,75 @@
+                             int flag)
+{
+	char new_term[1000] = "refs/bisect/";
+	strcat(new_term, term);
```
Please use strbuf here. (I am repeating so you don't forget.)

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -23,9 +79,22 @@ int cmd_bisect__helper(int argc, const char **argv, const char *prefix)
+			return c;
+		}
+		else
+			die(_("--check-term-format should be followed by exactly 2 arguments."));
```
Maybe something like:

```
	case CHECK_TERM_FMT:
		if (argc != 2)
			die(_("--check-term-format should be followed by exactly 2 arguments.")); 
		c = check_term_format(argv[0], argv[1], 0);
		printf("exit code: %d\n", c);
		return c;
```

chriscool replied:
Or better:

```
	case CHECK_TERM_FMT:
		if (argc != 2)
			die(_("--check-term-format should be followed by exactly 2 arguments.")); 
		return check_term_format(argv[0], argv[1], 0);
```

And if you really want the printf() you can add it in check_term_format().

chriscool commented on git-bisect.sh:
```diff
@@ -564,15 +564,29 @@ write_terms () {
+	echo "THis is PurE shit"
 	case "$term" in
 	help|start|terms|skip|next|reset|visualize|replay|log|run)
 		die "$(eval_gettext "can't use the builtin command '\$term' as a term")"
```
`die()` in shell writes on stderr, right?

Above in C the same text is written by:

```
printf("can't use the builtin command '%s' as a term\n", term);
```

but `printf()` writes on stdout, not stderr, so the C code is not doing the same thing as what the shell code did...

pranitbauva1997 replied:
die() in shell is apparently not writing to stderr. This is the problem I faced in commit no. 3 which is fixed in commit no. 4

That's why I wanted to clarify with you. Without the commit no. 4, 2 tests (no. 46 and 47) are failing. On a close look at those tests, you can clearly see that the stderr output is redirected to the file and then compared with a standard file. As a matter of fact, when you are using the shell function, check_term_format(), it runs fine which means that the string inside die() is not printed to stderr but it is still shown in the verbose output of running the test. I think the implementation of die() in shell is incorrect with git. I need you to verify all my observations once.

chriscool replied:
From some tests, it seems to me that `die` in shell is indeed writing on stderr:

```
> git bisect start --term-good=run --term-bad=skip >text
can't use the builtin command 'skip' as a term
> git bisect start --term-good=run --term-bad=skip >text 2>&1
>
```

chriscool replied:
You can also try:

```
> git bisect start --term-good=run --term-bad=skip 2>stderr_output
> cat stderr_output 
can't use the builtin command 'skip' as a term
```

chriscool commented:
In "git-sh-setup" there is:

```
die () {
	die_with_status 1 "$@"
}

die_with_status () {
	status=$1
	shift
	printf >&2 '%s\n' "$*"
	exit "$status"
}
```

So die() is printing on file descriptor 2 which is stderr.

pranitbauva1997 commented:
@chriscool See lines from 628 to 645 of [t/t6030](https://github.com/git/git/blob/master/t/t6030-bisect-porcelain.sh). In these tests it seems that die() of shell writes to stdout instead of stderr and my implementation in C (using die()) writes it to stderr. This can be easily noted from the verbose output. Thus there is a mis match in both the file (one being the template ie. `expected.ok`)

chriscool commented:
Here are the lines:

```
echo "" > expected.ok
cat > expected.missing-tree.default <<EOF
fatal: unable to read tree 39f7e61a724187ab767d2e08442d9b6b9dab587d
EOF

test_expect_success 'bisect fails if tree is broken on start commit' '
	git bisect reset &&
	test_must_fail git bisect start BROKEN_HASH7 BROKEN_HASH4 2>error.txt &&
	test_cmp expected.missing-tree.default error.txt
'

test_expect_success 'bisect fails if tree is broken on trial commit' '
	git bisect reset &&
	test_must_fail git bisect start BROKEN_HASH9 BROKEN_HASH4 2>error.txt &&
	git reset --hard broken &&
	git checkout broken &&
	test_cmp expected.missing-tree.default error.txt
'
```
The "2>error.txt" at the end of the `test_must_fail git bisect start ...` lines means that stderr is redirected to error.txt.

And the `test_cmp ...` lines compare the content of error.txt with the content of expected.missing-tree.default.

If you use `error("stuff")` in C it will produce "error: stuff" on stderr, so as expected.missing-tree.default contains an error that starts with "fatal:" and not "error:", there will be a test failure, but it is ok to change the test to make the error message in expected.missing-tree.default starts with "error:" instead of "fatal:".

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -33,10 +33,11 @@ static int one_of(const char *term, ...)
-	strcat(new_term, term);
+	struct strbuf new_term = STRBUF_INIT;
+	strbuf_addstr(&new_term, "refs/bisect/");
+	strbuf_addstr(&new_term, term);
```
To save one line, you could use:

```
struct strbuf new_term = STRBUF_INIT;
strbuf_addf(&new_term, "refs/bisect/%s", term);
```

pranitbauva1997 replied:
Thanks!

pranitbauva1997 commented:
@chriscool I think I wasn't clear in my description. Here is what I get as error when I run the tests with commit no 3. The problem is before there used to be **nothing** (ignore the tree for a while) written to the file. When I use error("can't change the meaning of the term '' "), it writes "error: can't change the meaning of the term '' " in the file while initially there used to be nothing written. Please try running the tests locally and compare the output with verbose to gain a proper perspective. The reason it took me a lot of while while writing all this is because I was trying out **every** permutation and combination to make it work and the current version is the only one which is working (passing the test suite).
I have reproduced the verbose output at commit no. 3
```
TERM_BAD: bad
TERM_GOOD: good
exit code: -1
exit code: -1
It does not come here
Bisecting: 0 revisions left to test after this (roughly 1 step)
--- expected.missing-tree.default	2016-05-03 13:47:37.276492026 +0000
+++ error.txt	2016-05-03 13:47:37.332493410 +0000
@@ -1 +1,5 @@
+error: can't change the meaning of the term 'bad'.fuck
+
+error: can't change the meaning of the term 'good'.fuck
+
 fatal: unable to read tree 39f7e61a724187ab767d2e08442d9b6b9dab587d
not ok 46 - bisect fails if tree is broken on start commit
#	
#		git bisect reset &&
#		test_must_fail git bisect start BROKEN_HASH7 BROKEN_HASH4 2>error.txt &&
#		test_cmp expected.missing-tree.default error.txt
#	

expecting success: 
	git bisect reset &&
	test_must_fail git bisect start BROKEN_HASH9 BROKEN_HASH4 2>error.txt &&
	git reset --hard broken &&
	git checkout broken &&
	test_cmp expected.missing-tree.default error.txt

We are not bisecting.
TERM_BAD: bad
TERM_GOOD: good
exit code: -1
exit code: -1
It does not come here
Bisecting: 2 revisions left to test after this (roughly 1 step)
HEAD is now at 508db08 9(broken): Remove missing file
--- expected.missing-tree.default	2016-05-03 13:47:37.276492026 +0000
+++ error.txt	2016-05-03 13:47:37.400495092 +0000
@@ -1 +1,5 @@
+error: can't change the meaning of the term 'bad'.fuck
+
+error: can't change the meaning of the term 'good'.fuck
+
 fatal: unable to read tree 39f7e61a724187ab767d2e08442d9b6b9dab587d
not ok 47 - bisect fails if tree is broken on trial commit
#	
#		git bisect reset &&
#		test_must_fail git bisect start BROKEN_HASH9 BROKEN_HASH4 2>error.txt &&
#		git reset --hard broken &&
#		git checkout broken &&
#		test_cmp expected.missing-tree.default error.txt
#	
```

This clearly says that the new implementation is introducing some lines which the previous one did not.

chriscool commented:
This shows that check_term_format() in C fails when it should not fail.

And I think it fails because in check_term_format() there is:

```
+	else if ((one_of(term, "bad", "new", NULL) && !strcmp(orig_term, "bad")) ||
+	    (one_of(term, "good", "old", NULL) && !strcmp(orig_term, "good"))){
+		printf("can't change the meaning of the term '%s'.fuck\n", term);
+		return -1;
+	}
```

And there should not be a `!` in front of `strcmp(orig_term, "bad")` and of `strcmp(orig_term, "good")`.

It should not fail if term == orig_term == "bad" and it should not fail if term == orig_term == "good".

`!strcmp(orig_term, "bad")` means `orig_term = "bad"` in shell, but what the shell tests is `orig_term != "bad"`.

pranitbauva1997 commented:
@chriscool That's too silly of me! Thanks a lot for your help. I was stuck here for ages. Will now clean the dirty stuff and finalize to send it to the mailing list. Also will start with another function 'bisect_log()" after this. I am currently searching about how is file handling done (the API for it). Probably you could speed up the process by directly pointing me to the code which uses it properly.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -42,8 +42,7 @@ static int check_term_format(const char *term, const char *orig_term,
-		printf("can't use the builtin command '%s' as a term\n", term);
-		return -1;
+		return error("can't use the builtin command '%s' as a term\n", term);
 	}
```
Here you can remove the braces since there is only one statement left.

pranitbauva1997 replied:
Sure!

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -53,10 +52,9 @@ static int check_term_format(const char *term, const char *orig_term,
+	else if ((one_of(term, "bad", "new", NULL) && strcmp(orig_term, "bad")) ||
+	    (one_of(term, "good", "old", NULL) && strcmp(orig_term, "good"))){
+		return error("can't change the meaning of the term '%s'.fuck\n", term);
 	}
```
Same here.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -15,6 +15,11 @@ enum sub_commands {
 };
 
+/*
+ * To check whether the string `term` belongs to the set of
```
You could start with just "Check ..."

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -15,6 +15,11 @@ enum sub_commands {
+/*
+ * To check whether the string `term` belongs to the set of
+ * strings included in the variable arguments to make the code
+ * look clean and avoid having long lines in the `if` statement.
```
I think you can remove "to make the code look clean and avoid having long lines in the `if` statement."
That could go in the commit message instead.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -81,10 +86,9 @@ int cmd_bisect__helper(int argc, const char **argv, const char *prefix)
-		else
+		if (argc != 2)
 			die(_("--check-term-format should be followed by exactly 2 arguments."));
+		return check_term_format(argv[0], argv[1], 0);
```
Good!

chriscool commented on git-bisect.sh:
```diff
@@ -564,29 +564,21 @@ write_terms () {
 	if [ $? -eq 128 ]; then
-		die "bye"
+		die "$(eval_gettext "there is a problem with the bad term")"
 	fi
```
Again I don't understand why the following is not enough:

```
git bisect--helper --check-term-format "$TERM_BAD" bad || exit
```

chriscool replied:
Why do you need this instead:

```
 	if [ $? -eq 128 ]; then
-		die "bye"
+		die "$(eval_gettext "there is a problem with the bad term")"
 	fi
```
?

chriscool commented on git-bisect.sh:
```diff
@@ -564,29 +564,21 @@ write_terms () {
 	if [ $? -eq 128 ]; then
-		die "bye"
+		die "$(eval_gettext "there is a problem with the good term")"
 	fi
```
Same here.

chriscool commented on git-bisect.sh:
```diff
@@ -564,29 +564,21 @@ write_terms () {
-	echo "THis is PurE shit"
 	case "$term" in
 	help|start|terms|skip|next|reset|visualize|replay|log|run)
 		die "$(eval_gettext "can't use the builtin command '\$term' as a term")"
```
Please don't forget to remove check_term_format() in "git-bisect.sh".

chriscool commented on git-bisect.sh:
```diff
@@ -564,8 +564,16 @@ write_terms () {
+	git bisect--helper --check-term-format "$TERM_GOOD" good
+	if [ $? -eq 128 ]; then
+		die "$(eval_gettext "there is a problem with the good term")"
+	fi
```
Please look at my previous comments about the above.

Also it looks like you didn't remove the check_term_format() function written in shell.

pranitbauva1997 replied:
I have mentioned this in my commit message that I don't plan to remove it now. After the whole rewriting exercise is finished, I can remove it. Do you want me to go the other way round? I have now used simple exit as you have said. I was previously trying with `exit 0`. Could you explain the difference between `|| exit` and `|| exit 0`

chriscool replied:
`foo || exit` is the same thing as `foo || exit $?`. That means that the exit code will be the exit code returned by foo.

`|| exit 0` is wrong here because it would mean success.

pranitbauva1997 commented:
@chriscool @larsxschneider I have now finished with this PR

pranitbauva1997 commented:
@chriscool I saw some work by Stefan Beller and he has removed the unnecessary parts so its better even I do it so. Thanks!

chriscool commented:
When you talk about other work (like Stefan Beller's work), it is better if you can add a link so we can easily take a look.

Otherwise the code looks very good now. 

Though the commit message could be improved. For example I am not sure that the second paragraph that starts with "When you call die() in the C implementation..." is very useful.

Also if you want to credit Lars and me, the best is probably to use "Mentored-by:" rather than something else.

chriscool commented:
About the commit message, I think something like:

"This reimplements the `check_term_format` shell function in C and adds a `--check-term-format` subcommand to `git bisect--helper` to call it from git-bisect.sh."

pranitbauva1997 commented:
@chriscool When mentioning about others work, I will henceforth make sure to include some links or commit ids. BTW the commit ids I was talking about are 74703a1, 0ea306e, ee8838d .

I will use mentored by from now.

Your commit message looks nice to me.

Thanks for your explanation about exit 0. Now it became clearer as to why the tests were failing.

chriscool commented:
It looks great to me now, except that you could also add a "Mentored-by:" for Lars.

I think, as long as we are both mentoring you (which will probably mean until the end of the GSoC), it is worth having both of us on all your GSoC related patches even if one, or both, of us doesn't comment much on some of the patches.

For example if three years from now, someone wants to reach people involved in the patches in this series, it could be useful to know that Lars was involved.

pranitbauva1997 commented:
@chriscool Sure! Will add it.

chriscool commented:
Great, thanks!
After that I am ok with you sending the series to the mailing list, but please wait a few hours to give Lars a chance to comment.

Ooops, I just saw that you just sent it. Maybe next time wait a bit to give Lars a chance to comment.
Thanks anyway.

pranitbauva1997 commented:
@chriscool Sorry for that! I will take care about it from next time!

pranitbauva1997 commented:
@chriscool @larsxschneider Hey! I have updated according to the reviews on the mailing list. Please have a look. [This](http://thread.gmane.org/gmane.comp.version-control.git/293488) is the conversation on mailing list for an easier review. Please suggest any modifications (if any) in the commit message for the first commit.

Changes done:

- Stick with 'subcommand' in the commit message.
- Rename enum as 'subcommand' to make it singular.
- s/bug/BUG/g
- Drop _() in the default of switch case
- Use some suggestions for commit message no. 2 and also include why I am using subcommand
- Drop of the later part in the comment of one_of()
- Remove flag from check_term_format()
- Use return error() instead of die()
- Drop '\n' from the localization strings
- Include localization for a missed string
- A comment in function check_term_format() is re-wrapped to fit in 80 columns
- Use 'term' instead of 'ref'
- Reword the string in the default switch case
- I kept the commit message same as previous for commit no.1 except for the minor one

I will now start working on bisect_log() conversion. I will branch off from this branch only as for now, then I will rebase it on Junio's local branch for these patches once they hit the next branch. This is the most appropriate way right?

pranitbauva1997 commented:
@chriscool @larsxschneider Hey! I have updated according to the reviews on the mailing list. Please have a look. [This](http://thread.gmane.org/gmane.comp.version-control.git/293488) is the conversation on mailing list for an easier review. Please suggest any modifications (if any) in the commit message for the first commit.

Changes done:

- Stick with 'subcommand' in the commit message.
- Rename enum as 'subcommand' to make it singular.
- s/bug/BUG/g
- Drop _() in the default of switch case
- Use some suggestions for commit message no. 2 and also include why I am using subcommand
- Drop of the later part in the comment of one_of()
- Remove flag from check_term_format()
- Use return error() instead of die()
- Drop '\n' from the localization strings
- Include localization for a missed string
- A comment in function check_term_format() is re-wrapped to fit in 80 columns
- Use 'term' instead of 'ref'
- Reword the string in the default switch case
- I kept the commit message same as previous for commit no.1 except for the minor one

I will now start working on bisect_log() conversion. I will branch off from this branch only as for now, then I will rebase it on Junio's local branch for these patches once they hit the next branch. This is the most appropriate way right?

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -2,23 +2,74 @@
+	 */
+
+	if ((strcmp(orig_term, "bad") && one_of("bad", "new", NULL)) ||
+		 (strcmp(orig_term, "good") && one_of("good", "old", NULL)))
```
It looks like the `one_of()` calls are missing `term` as their first argument.

chriscool replied:
By the way this shows that it might be good to increase test coverage to catch something like this.

pranitbauva1997 replied:
I should add a test to improve this. Will do in a different patch as a standalone afterwards.

chriscool commented:
Maybe you could add Lars as a mentor in the first patch too. Thanks.

chriscool commented:
>  I will branch off from this branch only as for now, then I will rebase it on Junio's local branch for these patches once they hit the next branch. This is the most appropriate way right?

Yeah, you can do that.

pranitbauva1997 commented:
@chriscool I had previously added Lars as a mentor along with you. Thanks for pointing out missing "term". Have updated it!

chriscool commented:
About Lars, yeah sorry it was a mistake.
Now it looks good to me.

larsxschneider commented:
@pranitbauva1997 I am reading your patches and I try to give feedback where I can. However, Christian is mentoring you way better than I am and I feel a bit bad if I am mentioned in the same way as he is in the commit messages. Therefore I would feel better if you would drop me until I manage to help you with one of my comments!

pranitbauva1997 commented:
@chriscool @larsxschneider I have updated the code according to the reviews. I have also included the write_terms() function. [Here](http://thread.gmane.org/gmane.comp.version-control.git/293785) is the discussion of the reviews from the mailing list. I have also incorporated the reviews on github. I have not used christian's suggestion of using `|| exit` as it breaks test no. 43 and 44

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -10,11 +10,11 @@ static const char * const git_bisect_helper_usage[] = {
 int cmd_bisect__helper(int argc, const char **argv, const char *prefix)
 {
-	int next_all = 0;
+	enum { NEXT_ALL = 1 } cmdmode = 0;
```
You could even do something like:
```
	enum {
		MODE_UNSPECIFIED = 0,
		MODE_NEXT_ALL
	} cmdmode = MODE_UNSPECIFIED;
```

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -23,9 +23,14 @@ int cmd_bisect__helper(int argc, const char **argv, const char *prefix)
+	case NEXT_ALL:
+		return bisect_next_all(prefix, no_checkout);
+	default:
+		die("BUG: unknown subcommand '%d'", cmdmode);
```
Now that the variable is called "cmdmode", please use:

```
		die("BUG: invalid cmdmode %d", (int)cmdmode);
```

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -56,18 +56,40 @@ static int check_term_format(const char *term, const char *orig_term)
+		fclose(fp);
+		return -1;
+	}
+	fclose(fp);
```
It looks like you also need to release the 'content' strbuf.

chriscool commented:
Could you show how the tests are broken by running them with -v and posting the output? Thanks!

chriscool commented:
@larsxschneider don't feel bad about being mentioned as a mentor in the commit. I will take some vacations at some points in July and August and then you will have more work. (Also I tend to help the student more at the beginning and less at the end of the GSoC.)

In order not to often burden Pranit with adding one or removing the other, it is simply better to just ask him to put both of us all the time during all the GSoC.

pranitbauva1997 commented:
@chriscool I have updated by releasing the strbuf memory. [Here](http://paste.ubuntu.com/16357011/) for the tests with `|| exit`.
Note: Test no. 43 and 44 are failing.

I will also investigate further as to why they are failing

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -56,18 +56,44 @@ static int check_term_format(const char *term, const char *orig_term)
+	fp = fopen(".git/BISECT_TERMS", "w");
+	if (!fp){
+		strbuf_release(&content);
+		die(_("could not open the file to read terms"));
```
Is there a reason why die() is used instead of error() here?

pranitbauva1997 replied:
@chriscool Well I don't think not being to open a file is a quite exceptional situation.

builtin/am.c uses return error(), builtin/fsck.c, builtin/submodule--helper.c and many more uses die_errno().

Quick question: Why do they use die_errno() rather than die(). What's the basic difference?

chriscool replied:
After a file related operation the errno variable contains an error number and it's worth it to print the related error, as it explains why the operation failed.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -56,18 +56,44 @@ static int check_term_format(const char *term, const char *orig_term)
+	}
+	fclose(fp);
+	strbuf_release(&content);
+	return 0;
```
you could do something like:

```
res = strbuf_write(&content, fp);
/* cleanup resources */
...
return (res < 0) ? -1 : 0;
```

pranitbauva1997 replied:
@chriscool This seems a lot better! Thanks!

chriscool commented:
The failures in test no. 43 and 44 seem related to the fact that bisect is run in bare repo.
So yeah it could be interesting to investigate why they are failing.

pranitbauva1997 commented:
@chriscool @larsxschneider I have made changes according to the recent reviews. Please have a look at it. I have removed the bisect_voc() function and will keep it on hold till its actually required.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -56,18 +58,38 @@ static int check_term_format(const char *term, const char *orig_term)
+	fp = fopen(git_path_bisect_write_terms(), "w");
+	if (!fp) {
+		return error(_("could not open the file to write terms"));
+	}
```
Please remove the useless braces.
You could also add a blank line after the `return error(...);`. Or if you prefer remove blanks line after the return statements above.
The best is to look at what other functions in the file are doing regarding blank lines and to do the same.

chriscool commented:
I think you can also remove the last patch about `git bisect log` as it is also not really related.

It looks good to me except the braces and blank line nits and also except that there should probably be `|| exit` after some calls...

pranitbauva1997 commented:
@chriscool I have removed the braces and added an extra line (most of the occurrences are in the same fashion).

Should I send bisect_log() in another patch?

chriscool commented:
I think bisect_log() is not worth sending now. Later it will be useful to finish the rewrite, but for now I doubt that Junio would merge it. So just keep it and focus on other things like investigating the `|| exit` issue. Thanks!

pranitbauva1997 commented:
@chriscool Sure! Thanks! I will send the patch now

pranitbauva1997 commented:
@chriscool I noticed something really weird. On this branch, I first put an `|| exit` and the test suite passed perfectly. Seeing this behavior, I also put `|| exit` on the master branch and the test suite still ran perfectly. Then I thought that it might have been because I used die() before and now I am using error(), so I switched back to die() and still the test suite passed. Then finally I picked up my patch v6[1] and applied it on master branch and then made the change `|| exit` and thus the test suite finally failed at t6030 test no 43 and 44.

This is something really unexpected. I am still investigating on this. Will keep you updated.

[1]: http://thread.gmane.org/gmane.comp.version-control.git/294388

pranitbauva1997 commented:
@chriscool @larsxschneider I got the reason why its failing. It is because ".git" was hardcoded. That's why it used to return an error but it wouldn't exit.

I think it would be most appropriate to incorporate the patch in the previous one(write-terms branch) and not send a patch to add `|| exit` (from master branch) because it would unnecessary clutter. What are your opinions?

Here is a verbose output[1] of running the test on the branch fail-write-terms[2].

[1]: http://paste.ubuntu.com/16635772/
[2]: https://github.com/pranitbauva1997/git/tree/fail-write-terms

chriscool commented:
Great that you found the reason why it was failing!
And yeah, I agree it should be integrated into the previous patch.
